File: programmers/lv2_review/lv_2_19.py
# 위장


"""
def solution(clothes):
    def dfs_left(s, idx):
        if idx == half:
            left.append(s)
            return
        dfs_left(s, idx + 1)
        dfs_left(s * values[idx], idx + 1)

    def dfs_right(s, idx):
        if idx == len(values):
            right.append(s)
            return
        dfs_right(s, idx + 1)
        dfs_right(s * values[idx], idx + 1)

    clothes_dic = {}
    for _, v in clothes:
        clothes_dic[v] = clothes_dic.get(v, 0) + 1
    values = list(clothes_dic.values())
    half = len(values) // 2
    left, right = [], []
    dfs_left(1, 0)
    dfs_right(1, half)
    answer = sum(left) + sum(right) - 2
    for i in range(1, len(left)):
        s = 0
        for j in range(1, len(right)):
            s += left[i] * right[j]
        answer += s
    return answer



"""


def solution(clothes):
    categories = {}
    for _, c in clothes:
        if c not in categories:
            categories[c] = 2
        else:
            categories[c] += 1

    cnt = 1
    values = list(categories.values())
    for i in range(len(values)):
        cnt *= values[i]
    cnt -= 1
    return cnt


clothes = [
    ["yellow_hat", "headgear"],
    ["blue_sunglasses", "eyewear"],
    ["green_turban", "headgear"],
]
print(solution(clothes))

File: programmers/lv2_review/lv2_19.js
const INF = 987654321;
const dy = [-1, 1, 0, 0];
const dx = [0, 0, -1, 1];

function solution(maps) {
  const q = [];
  const visited = Array.from({ length: maps.length }, () =>
    Array(maps[0].length).fill(INF)
  );
  visited[0][0] = 1;
  q.push([0, 0, 1]);
  while (q.length) {
    const [y, x, cnt] = q.shift();
    for (let i = 0; i < 4; i++) {
      ny = y + dy[i];
      nx = x + dx[i];
      if (ny < 0 || ny >= maps.length || nx < 0 || nx >= maps[0].length)
        continue;
      if (maps[ny][nx] === 0) continue;
      if (visited[ny][nx] > cnt + 1) {
        visited[ny][nx] = cnt + 1;
        q.push([ny, nx, cnt + 1]);
      }
    }
  }
  const answer = visited[maps.length - 1][maps[0].length - 1];
  return answer < INF ? answer : -1;
}

const maps = [
  [1, 0, 1, 1, 1],
  [1, 0, 1, 0, 1],
  [1, 0, 1, 1, 1],
  [1, 1, 1, 0, 0],
  [0, 0, 0, 0, 1],
];
console.log(bfs(maps));
